Doc revised - Groth16 page (#7459)
iotaledger/iota: docs/content/developer/cryptography/on-chain/groth16.mdx

There are high-level languages for expressing these statements, such as [Circom](https://docs.circom.io), used in the following example.

Groth16 requires a trusted setup for each circuit to generate the verification key. The API is not pinning any particular verification key and each user can generate their own parameters or use an existing verification to their apps. 

> Please refer to this page for ceremony setup [Proving circuits with ZK](https://docs.circom.io/getting-started/proving-circuits/), starting from phase 1. Also, instead of using the one from local, you can also use the one from [powersoftau](https://github.com/iden3/snarkjs?tab=readme-ov-file#7-prepare-phase-2), which is a common good and can be used as a starting point of phase 2.

Reviewer: > from phase 1

The page referred does not have "phase 1", it has "Powers of Tau" phase instead, maybe refer to it this way?

> Please refer to this page for ceremony setup [Proving circuits with ZK](https://docs.circom.io/getting-started/proving-circuits/), starting from phase 1. Also, instead of using the one from local, you can also use the one from [powersoftau](https://github.com/iden3/snarkjs?tab=readme-ov-file#7-prepare-phase-2), which is a common good and can be used as a starting point of phase 2.

Reviewer: > the one from local

What does "the one from local" refer to? Do you mean, instead of generating powers of tau locally (because it is a computationally intense operation) users can fetch pregenerated powers of tau from iden3 github page?

> Please refer to this page for ceremony setup [Proving circuits with ZK](https://docs.circom.io/getting-started/proving-circuits/), starting from phase 1. Also, instead of using the one from local, you can also use the one from [powersoftau](https://github.com/iden3/snarkjs?tab=readme-ov-file#7-prepare-phase-2), which is a common good and can be used as a starting point of phase 2.

Reviewer: > a common good

That's very subjective. Maybe just say that iden3 github page has pregenerated powers of tau and let the user decide whether to trust/use it or not. After all, there's no way to verify those were generated honestly.

```bash

> wget https://hermez.s3-eu-west-1.amazonaws.com/powersOfTau28_hez_final_12.ptau      -O pot12_final.ptau

Reviewer: ```suggestion
wget https://hermez.s3-eu-west-1.amazonaws.com/powersOfTau28_hez_final_12.ptau -O pot12_final.ptau
```

> node generate_witness.js circuit.wasm input.json witness.wtns

Reviewer: You should either explain better what's going on here, including commands and arguments, otherwise it will be unhelpful and confusing to the users. Or you can refer to snarkjs docs, it's their job to explain how to use `snarkjs` and maintain the docs.

snarkjs powersoftau new bn128 12 pot12_0000.ptau -v
snarkjs powersoftau contribute pot12_0000.ptau pot12_0001.ptau --name="First contribution" -v
snarkjs powersoftau prepare phase2 pot12_0001.ptau pot12_final.ptau -v
snarkjs groth16 setup ../circuit.r1cs pot12_final.ptau circuit_0000.zkey
snarkjs zkey contribute circuit.zkey circuit_0001.zkey --name="1st Contributor Name" -v
snarkjs zkey export verificationkey circuit_0001.zkey verification_key.json
```
Which completes the export of verification key.

## Usage

The following example demonstrates how to create a Groth16 proof from a statement written in Circom and then verify it using the IOTA Move API. The API currently supports up to eight public inputs.

Add the following dependencies in your Cargo.toml:
```
ark-bn254 = "0.5.0"
ark-circom = "0.5.0"
ark-groth16 = "0.5.0"
ark-serialize = "0.5.0"
ark-snark = "0.5.1"
hex = "0.4.3"

> rand = "0.8.5"

Reviewer 2: Perhaps it would be better to use `ark_std` which also [reexports](https://docs.rs/ark-std/0.5.0/ark_std/index.html#reexport.rand) rand 0.8.5 instead of pulling rand yourself

Reviewer 2: Just a suggestion

Author: Updated. My oversight on ark-std's reexport.

That's a good catch! Thanks sir @marc2332 .

> ark-std = "0.5.0"

Reviewer: `ark-std` looks nicer one line above)

### Generate proof

> To generate a proof verifiable in IOTA, you need to generate a witness. This example uses Arkworks' [ark-circom](https://github.com/gakonst/ark-circom) Rust library. The code constructs a witness for the circuit and generates a proof for it for a given input. Finally, it verifies that the proof is correct. Be advised that the curve in the below exampl does not satisfy the 128-bit security level anymore. We suggest you to use the curve BLS12-381 G1 curve instead.

Reviewer: ```suggestion
To generate a proof verifiable in IOTA, you need to generate a witness. This example uses Arkworks' [ark-circom](https://github.com/gakonst/ark-circom) Rust library. The code constructs a witness for the circuit and generates a proof for it for a given input. Finally, it verifies that the proof is correct. Be advised that the curve BN254 in the example below does not satisfy the 128-bit security level anymore. We suggest you use BLS12-381 curve instead.
```

    // WARNING: The code below is just for debugging, and should instead use a verification key generated from a trusted setup.
    // See for example https://docs.circom.io/getting-started/proving-circuits/#powers-of-tau.

>     let mut rng = SeedableRng::from_seed([0; 32]);

Reviewer: ```suggestion
    let mut rng = StdRng::from_seed([0; 32]);
```

>     let mut rng = StdRng = SeedableRng::from_seed([0; 32]);

Reviewer 2: ```suggestion
    let mut rng = SeedableRng::from_seed([0; 32]);
```

    let proof = Groth16::<Bn254>::prove(&params, circom, &mut rng).unwrap();

    // Check that the proof is valid
    let pvk = prepare_verifying_key(&params.vk);
    let verified = Groth16::<Bn254>::verify_with_processed_vk(&pvk, &inputs, &proof).unwrap();
    assert!(verified);
}
```
The proof shows that an input (7) which, when hashed with the Poseidon hash function, gives a certain output (which in this case is `inputs[0].to_string() = 7061949393491957813657776856458368574501817871421526214197139795307327923534`).

You can also use snarkjs to generate and export proof yourself to observe the proof body.
```bash
snarkjs groth16 prove circuit_0001.zkey witness.wtns proof.json public.json
```

### Verification in IOTA

> The API in IOTA for verifying a proof expects a special processed verification key, where only a subset of the values are used. Ideally, computation for this prepared verification key happens only once per circuit. You can perform this processing using the `iota::groth16::prepare_verifying_key` method of the IOTA Move API with a serialization of the `params.vk` value used previously. Remember that you should not use this in a production environment, as you should grab a safer entropy resource i.e., from a ceremony.

Reviewer: > Remember that you should not use this in a production environment, as you should grab a safer entropy resource i.e., from a ceremony.

Weird phrasing. What exactly "this" should not be used in production? Verification key processing with `iota::groth16::prepare_verifying_key`? Why entropy "resource" here? The phrase doesn't make sense to me, it would be even more confusing to users.

> The API in IOTA for verifying a proof expects a special processed verification key, where only a subset of the values are used. Ideally, computation for this prepared verification key happens only once per circuit. You can perform this processing using the `iota::groth16::prepare_verifying_key` method of the IOTA Move API with a serialization of the `params.vk` value used previously. Remember that you should not use this in the production environment, as you should grab a safer entropy resource i.e., from a ceremony.

Reviewer 2: ```suggestion
The API in IOTA for verifying a proof expects a special processed verification key, where only a subset of the values are used. Ideally, computation for this prepared verification key happens only once per circuit. You can perform this processing using the `iota::groth16::prepare_verifying_key` method of the IOTA Move API with a serialization of the `params.vk` value used previously. Remember that you should not use this in a production environment, as you should grab a safer entropy resource i.e., from a ceremony.
```

To verify a proof, you also need two more inputs, `public_inputs_bytes` and `proof_points_bytes`, which contain the public inputs and the proof respectively. These are serializations of the `inputs` and `proof` values from the previous example, which you can compute in Rust as follows:

```rust
// Print verifying key
let mut vk_bytes = Vec::new();
params.vk.serialize_compressed(&mut vk_bytes).unwrap();
println!("Verifying key: {}", hex::encode(vk_bytes));

// Print public inputs. Note that they are concatenated.
let mut public_inputs_bytes = Vec::new();
for i in 0..inputs.len() { // if there is more than one public input, serialize one by one
    inputs[i].serialize_compressed(&mut inputs_bytes).unwrap();
}
println!("Public inputs: {}", hex::encode(public_inputs_serialized));

// Print proof
let mut proof_points_bytes = Vec::new();
proof.serialize_compressed(&mut proof_points_bytes).unwrap();
println!("Proof: {}", hex::encode(proof_serialized));
```

You can also use snarkjs to verify the generated and exported proof.
```bash
snarkjs groth16 verify verification_key.json public.json proof.json
```

> The following example smart contract prepares a verification key and verifies the corresponding proof. This example uses the BN254 elliptic curve construction, which is given as the first parameter to the `prepare_verifying_key` and `verify_groth16_proof` functions. You can use the [`bls12381`](https://docs.iota.org/references/framework/iota/groth16#iota_groth16_bls12381) function instead, as bn254 is not considered safe enough anymore.

Reviewer: - Please use standard curve names BLS12-381 and BN254.
- BLS12-381 is a curve, not a function.
- This security concern has already been mentioned above, no need to state that again IMO.
- Why not rewrite the example using BLS12-381 directly, and say that users can use a less secure curve BN254?
